lib/plugin/opsgenie/client.go

```diff
@@ -0,0 +1,281 @@
+{{if .RequestReason}}Reason: {{.RequestReason}}{{end}}
+{{if .RequestLink}}To approve or deny the request, proceed to {{.RequestLink}}{{end}}
+`,
+))
+var reviewNoteTemplate = template.Must(template.New("review note").Parse(
+	`{{.Author}} reviewed the request at {{.Created.Format .TimeFormat}}.
+Resolution: {{.ProposedState}}.
+{{if .Reason}}Reason: {{.Reason}}.{{end}}`,
+))
+var resolutionNoteTemplate = template.Must(template.New("resolution note").Parse(
+	`Access request has been {{.Resolution}}
+{{if .ResolveReason}}Reason: {{.ResolveReason}}{{end}}`,
+))
+
+// Client is a wrapper around resty.Client.
+type Client struct {
+	ClientConfig
+
+	client *resty.Client
+}
+
+type ClientConfig struct {
+	// APIKey is the API key for Opsgenie
+	APIKey string
+	// APIEndpoint is the endpoint for the Opsgenie API
+	APIEndpoint string
+	// DefaultSchedules are the default on-call schedules to check for auto approval
+	DefaultSchedules []string
+	// Priority is the priority alerts are to be created with
+	Priority string
+
+	// WebProxyURL is the Teleport address used when building the bodies of the alerts
+	// allowing links to the access requests to be built
+	WebProxyURL *url.URL
+	// ClusterName is the name of the Teleport cluster
+	ClusterName string
+}
+
+// NewClient creates a new Opsgenie client for managing alerts.
+func NewClient(conf ClientConfig) (*Client, error) {
+	client := resty.NewWithClient(defaults.Config().HTTPClient)
+	client.SetHeader("Authorization", "GenieKey "+conf.APIKey)
+	client.SetHostURL(conf.APIEndpoint)
+	return &Client{
+		client:       client,
+		ClientConfig: conf,
+	}, nil
+}
+
+func errWrapper(statusCode int) error {
+	switch statusCode {
+	case http.StatusForbidden:
+		return trace.AccessDenied("opsgenie API access denied: status code %v", statusCode)
+	case http.StatusRequestTimeout:
+		return trace.ConnectionProblem(trace.Errorf("status code %v", statusCode),
+			"connecting to opsgenie API")
+	}
+	return trace.Errorf("connecting to opsgenie API status code %v", statusCode)
+}
+
+// CreateAlert creates an opsgenie alert.
+func (og Client) CreateAlert(ctx context.Context, reqID string, reqData RequestData) (OpsgenieData, error) {
+	bodyDetails, err := buildAlertBody(og.WebProxyURL, reqID, reqData)
+	if err != nil {
+		return OpsgenieData{}, trace.Wrap(err)
+	}
+
+	body := AlertBody{
+		Message:     fmt.Sprintf("Access request from %s", reqData.User),
+		Alias:       fmt.Sprintf("%s/%s", alertKeyPrefix, reqID),
+		Description: bodyDetails,
+		Responders:  og.getResponders(reqData),
+		Priority:    og.Priority,
+	}
+
+	var result AlertResult
+	resp, err := og.client.NewRequest().
+		SetContext(ctx).
+		SetBody(body).
+		SetResult(&result).
+		Post("alerts")
+
+	if err != nil {
+		return OpsgenieData{}, trace.Wrap(err)
+	}
+	if resp.IsError() {
+		return OpsgenieData{}, errWrapper(resp.StatusCode())
+	}
+	defer resp.RawResponse.Body.Close()
```

Comment:
`defer` should happen before we check the HTTP status code. Here if `resp.IsError()` we never close the body and return immediately. (this applies to other HTTP calls as well)